### PR TITLE
fix(amplify_authenticator): allow multiple custom attributes in custom forms (v2)

### PR DESCRIPTION
**Pull request by Jordan-Nelson**
*Issue #, if available:*

*Description of changes:*
Alternate to https://github.com/aws-amplify/amplify-flutter/pull/1377 with a different public API

Example API

```dart
// custom attributes
signUpForm: SignUpForm.custom(
  includeSignUpAttributes: true,
  includeVerificationAttributes: true,
  fields: [
    ...SignUpForm.requiredFields,
    SignUpFormField.custom(
      title: 'Age',
      attributeKey: const CognitoUserAttributeKey.custom('age'),
    ),
    SignUpFormField.custom(
      title: 'Bio',
      attributeKey: const CognitoUserAttributeKey.custom('bio'),
    ),
  ],
),

// custom attributes, no runtime attributes
signUpForm: SignUpForm.custom(
  fields: [
    ...SignUpForm.requiredFields,
    SignUpFormField.custom(
      title: 'Age',
      attributeKey: const CognitoUserAttributeKey.custom('age'),
    ),
    SignUpFormField.custom(
      title: 'Bio',
      attributeKey: const CognitoUserAttributeKey.custom('bio'),
    ),
  ],
),

// password confirmation field removed
signUpForm: SignUpForm.custom(
  includeSignUpAttributes: true,
  includeVerificationAttributes: true,
  fields: [
    SignUpFormField.username(),
    SignUpFormField.password(),
  ],
),

// one additional attribute
signUpForm: SignUpForm.custom(
  fields: [
    ...SignUpForm.requiredFields,
    SignUpFormField.address(),
  ],
),

// additional attribute listed first
signUpForm: SignUpForm.custom(
  fields: [
    SignUpFormField.address(),
    ...SignUpForm.requiredFields,
  ],
),

// just verification
signUpForm: SignUpForm.custom(
  includeVerificationAttributes: true,
  fields: [
    ...SignUpForm.requiredFields,
    SignUpFormField.address(),
  ],
),

// just sign up attributes
signUpForm: SignUpForm.custom(
  includeSignUpAttributes: true,
  fields: [
    ...SignUpForm.requiredFields,
    SignUpFormField.address(),
  ],
),
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_authenticator/lib/src/widgets/component.dart [hunk 1]
```diff
@@ -50,7 +50,7 @@ abstract class StatelessAuthenticatorComponent extends StatelessWidget {
 /// {@template amplify_authenticator.authenticator_component}
 /// Base class for all stateful Authenticator components.
 /// {@endtemplate}
-abstract class AuthenticatorComponent<T extends AuthenticatorComponent<T>>
+abstract class AuthenticatorComponent<T extends StatefulWidget>
```
Comment (dnys1) [inline packages/amplify_authenticator/lib/src/widgets/component.dart:53]: Are these changes still necessary? Not seeing where it affects things

File: packages/amplify_authenticator/lib/src/widgets/form.dart [hunk 2]
```diff
@@ -262,47 +232,125 @@ class SignUpForm extends AuthenticatorForm {
...
+    ));
+  }
 }
 
 class _SignUpFormState extends AuthenticatorFormState<SignUpForm> {
   _SignUpFormState() : super();
 
   @override
-  List<SignUpFormField> runtimeFields(BuildContext context) {
+  List<SignUpFormField> get allFields {
```
Comment (dnys1) [inline packages/amplify_authenticator/lib/src/widgets/form.dart:308]: Should this logic be located here? Seems like it might be needed in other forms as well.

Comment (Jordan-Nelson) [reply packages/amplify_authenticator/lib/src/widgets/form.dart:308]: It isn't used in any other forms currently. It's possible that it could be used in other forms in the future, but I cannot think of a use case for it.

**Review by dnys1 — APPROVED** [review]

File: packages/amplify_authenticator/lib/src/widgets/component.dart (hunk 1 above)
Comment (Jordan-Nelson) [reply packages/amplify_authenticator/lib/src/widgets/component.dart:53]: Yes they are needed.

These changes are needed to make the changes to `AuthenticatorFormState`. Without those the changes to `AuthenticatorFormState`, `widget` within `_SignUpFormState` is of type `AuthenticatorForm` instead of the more specific `SignUpForm`.

**Comment by Jordan-Nelson** [comment]
closing in favor of https://github.com/aws-amplify/amplify-flutter/pull/1377
